File: agent_lifecycle_protocol/lifecycle.py
"""Lifecycle state machine and transition logic (Section 4).

Manages agent state transitions with precondition checks and hook points.
"""

from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional

from .schema import (
    AGENT_STATES,
    TERMINAL_STATES,
    AgentRecord,
    EpigeneticProfile,
    GeneticProfile,
    Initiator,
    LifecycleEvent,
    _now_iso,
    _uuid,
)
from .store import LifecycleStore


class LifecycleError(Exception):
    """Raised for invalid lifecycle transitions."""


# Valid transitions: from_state -> {to_state: [allowed_event_types]}
_TRANSITIONS: Dict[Optional[str], Dict[str, List[str]]] = {
    None: {
        "provisioning": ["genesis"],
    },
    "provisioning": {
        "active": ["activate"],
        "failed": ["fail"],
    },
    "active": {
        "suspended": ["suspend"],
        "migrating": ["begin_migration"],
        "deprecated": ["deprecate"],
        "active": ["retraining", "fork"],  # self-transitions
        "decommissioned": ["emergency_decommission"],
    },
    "suspended": {
        "active": ["resume"],
        "decommissioned": ["emergency_decommission"],
    },
    "migrating": {
        "active": ["complete_migration", "abort_migration"],
        "decommissioned": ["emergency_decommission"],
    },
    "deprecated": {
        "decommissioned": ["decommission"],
        "active": ["abort_succession"],
    },
}


@dataclass
class HookResult:
    """Result from a lifecycle hook execution."""
    success: bool = True
    error: str = ""
    data: Dict[str, Any] = field(default_factory=dict)


# Type for hook callbacks: (event, agent_record) -> HookResult
HookCallback = Callable[[LifecycleEvent, AgentRecord], HookResult]


class LifecycleManager:
    """Manages agent lifecycle state machine transitions.

    Provides genesis, activate, suspend, resume, deprecate, decommission,
    and emergency_decommission operations. Fork, migration, succession,
    and retraining are in their own modules but integrate through this
    state machine.
    """

    def __init__(
        self,
        store: Optional[LifecycleStore] = None,
        store_dir: str = ".alp",
    ) -> None:
        self.store = store or LifecycleStore(store_dir)
        self._pre_hooks: Dict[str, List[HookCallback]] = {}
        self._post_hooks: Dict[str, List[HookCallback]] = {}

    # -- Hook registration --

    def register_pre_hook(self, event_type: str, hook: HookCallback) -> None:
        self._pre_hooks.setdefault(event_type, []).append(hook)

    def register_post_hook(self, event_type: str, hook: HookCallback) -> None:
        self._post_hooks.setdefault(event_type, []).append(hook)

    def _run_pre_hooks(self, event: LifecycleEvent, agent: AgentRecord) -> None:
        for hook in self._pre_hooks.get(event.event_type, []):
            result = hook(event, agent)
            if not result.success:
                raise LifecycleError(
                    f"PreTransition hook failed for {event.event_type}: {result.error}"
                )

    def _run_post_hooks(self, event: LifecycleEvent, agent: AgentRecord) -> None:
        for hook in self._post_hooks.get(event.event_type, []):
            hook(event, agent)  # post-hooks cannot abort

    # -- Transition validation --

    def validate_transition(
        self, from_state: Optional[str], to_state: str, event_type: str
    ) -> None:
        allowed = _TRANSITIONS.get(from_state, {})
        if to_state not in allowed:
            raise LifecycleError(
                f"Invalid transition: {from_state} -> {to_state}"
            )
        if event_type not in allowed[to_state]:
            raise LifecycleError(
                f"Event type '{event_type}' not allowed for "
                f"{from_state} -> {to_state}"
            )

    def _apply_transition(
        self,
        agent: AgentRecord,
        event_type: str,
        to_state: str,
        initiator: Optional[Initiator] = None,
        details: Optional[Dict[str, Any]] = None,
    ) -> LifecycleEvent:
        from_state = agent.state
        self.validate_transition(from_state, to_state, event_type)

        event = LifecycleEvent(
            event_type=event_type,
            agent_id=agent.agent_id,
            agent_state_before=from_state,
            agent_state_after=to_state,
            initiator=initiator or Initiator(),
            details=details or {},
        )

        self._run_pre_hooks(event, agent)

        agent.state = to_state
        agent.updated_at = _now_iso()
        agent.event_history.append(event.event_id)

        event.compute_hash()
        self.store.append_event(event)
        self.store.save_agent(agent)

        self._run_post_hooks(event, agent)

        return event

    # -- Lifecycle operations --

    def genesis(
        self,
        agent_id: str,
        genetic_profile: Optional[GeneticProfile] = None,
        epigenetic_profile: Optional[EpigeneticProfile] = None,
        creator_id: str = "",
        creation_method: str = "manual",
        coc_chain_id: str = "",
    ) -> AgentRecord:
        """Create a new agent (Section 5.2). Returns agent in Provisioning."""
        existing = self.store.get_agent(agent_id)
        if existing is not None:
            raise LifecycleError(f"Agent already exists: {agent_id}")

        agent = AgentRecord(
            agent_id=agent_id,
            state="provisioning",
            genetic_profile=genetic_profile or GeneticProfile(),
            epigenetic_profile=epigenetic_profile or EpigeneticProfile(),
            coc_chain_id=coc_chain_id or f"coc-{agent_id}",
        )

        details = {
            "creation_method": creation_method,
            "genetic_profile": agent.genetic_profile.to_dict(),
            "epigenetic_profile": agent.epigenetic_profile.to_dict(),
            "identity": {
                "agent_id": agent_id,
                "coc_chain_id": agent.coc_chain_id,
            },
            "authorization": {
                "creator_id": creator_id,
            },
        }

        event = LifecycleEvent(
            event_type="genesis",
            agent_id=agent_id,
            agent_state_before=None,
            agent_state_after="provisioning",
            initiator=Initiator(type="human", id=creator_id),
            details=details,
        )

        self._run_pre_hooks(event, agent)

        agent.event_history.append(event.event_id)
        event.compute_hash()
        self.store.append_event(event)
        self.store.save_agent(agent)

        self._run_post_hooks(event, agent)

        return agent

    def activate(
        self,
        agent_id: str,
        initiator: Optional[Initiator] = None,
    ) -> AgentRecord:
        """Transition agent from Provisioning to Active (Section 4.2)."""
        agent = self._get_agent_or_raise(agent_id)
        self._apply_transition(agent, "activate", "active", initiator)
        return agent

    def suspend(
        self,
        agent_id: str,
        reason: str = "",
        initiator: Optional[Initiator] = None,
    ) -> AgentRecord:
        """Transition agent from Active to Suspended."""
        agent = self._get_agent_or_raise(agent_id)
        self._apply_transition(
            agent, "suspend", "suspended", initiator,
            details={"reason": reason},
        )
        return agent

    def resume(
        self,
        agent_id: str,
        initiator: Optional[Initiator] = None,
    ) -> AgentRecord:
        """Transition agent from Suspended to Active."""
        agent = self._get_agent_or_raise(agent_id)
        self._apply_transition(agent, "resume", "active", initiator)
        return agent

    def deprecate(
        self,
        agent_id: str,
        reason: str = "",
        successor_id: str = "",
        initiator: Optional[Initiator] = None,
    ) -> AgentRecord:
        """Transition agent from Active to Deprecated (Section 4.2)."""
        agent = self._get_agent_or_raise(agent_id)
        self._apply_transition(
            agent, "deprecate", "deprecated", initiator,
            details={"reason": reason, "successor_id": successor_id},
        )
        return agent

    def decommission(
        self,
        agent_id: str,
        reason: str = "end_of_life",
        estate_disposition: Optional[Dict[str, Any]] = None,
        initiator: Optional[Initiator] = None,
    ) -> AgentRecord:
        """Transition agent from Deprecated to Decommissioned (Section 5.7)."""
        agent = self._get_agent_or_raise(agent_id)
        self._apply_transition(
            agent, "decommission", "decommissioned", initiator,
            details={
                "reason": reason,
                "estate_disposition": estate_disposition or {},
            },
        )
        return agent

    def emergency_decommission(
        self,
        agent_id: str,
        reason: str = "compromised",
        initiator: Optional[Initiator] = None,
    ) -> AgentRecord:
        """Emergency decommission from any non-terminal state (Section 9.4)."""
        agent = self._get_agent_or_raise(agent_id)
        if agent.state in TERMINAL_STATES:
            raise LifecycleError(
                f"Cannot emergency-decommission agent in terminal state: {agent.state}"
            )

        event = LifecycleEvent(
            event_type="emergency_decommission",
            agent_id=agent_id,
            agent_state_before=agent.state,
            agent_state_after="decommissioned",
            initiator=initiator or Initiator(),
            details={"reason": reason, "emergency": True},
        )

        self._run_pre_hooks(event, agent)

        agent.state = "decommissioned"
        agent.updated_at = _now_iso()
        agent.event_history.append(event.event_id)

        event.compute_hash()
        self.store.append_event(event)
        self.store.save_agent(agent)

        self._run_post_hooks(event, agent)

        return agent

    def fail(
        self,
        agent_id: str,
        error: str = "",
        initiator: Optional[Initiator] = None,
    ) -> AgentRecord:
        """Transition agent from Provisioning to Failed."""
        agent = self._get_agent_or_raise(agent_id)
        self._apply_transition(
            agent, "fail", "failed", initiator,
            details={"error": error},
        )
        return agent

    def get_agent(self, agent_id: str) -> Optional[AgentRecord]:
        return self.store.get_agent(agent_id)

    def _get_agent_or_raise(self, agent_id: str) -> AgentRecord:
        agent = self.store.get_agent(agent_id)
        if agent is None:
            raise LifecycleError(f"Agent not found: {agent_id}")
        return agent

File: typescript/src/lifecycle.ts
import {
  TERMINAL_STATES,
  AgentRecord,
  EpigeneticProfile,
  GeneticProfile,
  Initiator,
  LifecycleEvent,
  nowIso,
} from "./types";
import { LifecycleStore } from "./store";

export class LifecycleError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "LifecycleError";
  }
}

export interface HookResult {
  success: boolean;
  error: string;
  data: Record<string, unknown>;
}

export type HookCallback = (
  event: LifecycleEvent,
  agent: AgentRecord,
) => HookResult;

const _TRANSITIONS: Record<
  string,
  Record<string, string[]>
> = {
  __none__: {
    provisioning: ["genesis"],
  },
  provisioning: {
    active: ["activate"],
    failed: ["fail"],
  },
  active: {
    suspended: ["suspend"],
    migrating: ["begin_migration"],
    deprecated: ["deprecate"],
    active: ["retraining", "fork"],
    decommissioned: ["emergency_decommission"],
  },
  suspended: {
    active: ["resume"],
    decommissioned: ["emergency_decommission"],
  },
  migrating: {
    active: ["complete_migration", "abort_migration"],
    decommissioned: ["emergency_decommission"],
  },
  deprecated: {
    decommissioned: ["decommission"],
    active: ["abort_succession"],
  },
};

export class LifecycleManager {
  readonly store: LifecycleStore;
  _preHooks: Record<string, HookCallback[]> = {};
  _postHooks: Record<string, HookCallback[]> = {};

  constructor(opts?: { store?: LifecycleStore; storeDir?: string }) {
    this.store = opts?.store ?? new LifecycleStore(opts?.storeDir ?? ".alp");
  }

  registerPreHook(eventType: string, hook: HookCallback): void {
    if (!this._preHooks[eventType]) this._preHooks[eventType] = [];
    this._preHooks[eventType].push(hook);
  }

  registerPostHook(eventType: string, hook: HookCallback): void {
    if (!this._postHooks[eventType]) this._postHooks[eventType] = [];
    this._postHooks[eventType].push(hook);
  }

  _runPreHooks(event: LifecycleEvent, agent: AgentRecord): void {
    for (const hook of this._preHooks[event.eventType] ?? []) {
      const result = hook(event, agent);
      if (!result.success) {
        throw new LifecycleError(
          `PreTransition hook failed for ${event.eventType}: ${result.error}`,
        );
      }
    }
  }

  _runPostHooks(event: LifecycleEvent, agent: AgentRecord): void {
    for (const hook of this._postHooks[event.eventType] ?? []) {
      hook(event, agent);
    }
  }

  validateTransition(
    fromState: string | null,
    toState: string,
    eventType: string,
  ): void {
    const key = fromState ?? "__none__";
    const allowed = _TRANSITIONS[key] ?? {};
    if (!(toState in allowed)) {
      throw new LifecycleError(
        `Invalid transition: ${fromState} -> ${toState}`,
      );
    }
    if (!allowed[toState].includes(eventType)) {
      throw new LifecycleError(
        `Event type '${eventType}' not allowed for ${fromState} -> ${toState}`,
      );
    }
  }

  private applyTransition(
    agent: AgentRecord,
    eventType: string,
    toState: string,
    initiator?: Initiator,
    details?: Record<string, unknown>,
  ): LifecycleEvent {
    const fromState = agent.state;
    this.validateTransition(fromState, toState, eventType);

    const event = new LifecycleEvent({
      eventType,
      agentId: agent.agentId,
      agentStateBefore: fromState,
      agentStateAfter: toState,
      initiator: initiator ?? new Initiator(),
      details: details ?? {},
    });

    this._runPreHooks(event, agent);

    agent.state = toState;
    agent.updatedAt = nowIso();
    agent.eventHistory.push(event.eventId);

    event.computeHash();
    this.store.appendEvent(event);
    this.store.saveAgent(agent);

    this._runPostHooks(event, agent);

    return event;
  }

  genesis(opts: {
    agentId: string;
    geneticProfile?: GeneticProfile;
    epigeneticProfile?: EpigeneticProfile;
    creatorId?: string;
    creationMethod?: string;
    cocChainId?: string;
  }): AgentRecord {
    const existing = this.store.getAgent(opts.agentId);
    if (existing !== null) {
      throw new LifecycleError(`Agent already exists: ${opts.agentId}`);
    }

    const agent = new AgentRecord({
      agentId: opts.agentId,
      state: "provisioning",
      geneticProfile: opts.geneticProfile ?? new GeneticProfile(),
      epigeneticProfile: opts.epigeneticProfile ?? new EpigeneticProfile(),
      cocChainId: opts.cocChainId ?? `coc-${opts.agentId}`,
    });

    const details: Record<string, unknown> = {
      creation_method: opts.creationMethod ?? "manual",
      genetic_profile: agent.geneticProfile.toDict(),
      epigenetic_profile: agent.epigeneticProfile.toDict(),
      identity: {
        agent_id: opts.agentId,
        coc_chain_id: agent.cocChainId,
      },
      authorization: {
        creator_id: opts.creatorId ?? "",
      },
    };

    const event = new LifecycleEvent({
      eventType: "genesis",
      agentId: opts.agentId,
      agentStateBefore: null,
      agentStateAfter: "provisioning",
      initiator: new Initiator({ type: "human", id: opts.creatorId ?? "" }),
      details,
    });

    this._runPreHooks(event, agent);

    agent.eventHistory.push(event.eventId);
    event.computeHash();
    this.store.appendEvent(event);
    this.store.saveAgent(agent);

    this._runPostHooks(event, agent);

    return agent;
  }

  activate(agentId: string, initiator?: Initiator): AgentRecord {
    const agent = this.getAgentOrThrow(agentId);
    this.applyTransition(agent, "activate", "active", initiator);
    return agent;
  }

  suspend(
    agentId: string,
    reason: string = "",
    initiator?: Initiator,
  ): AgentRecord {
    const agent = this.getAgentOrThrow(agentId);
    this.applyTransition(agent, "suspend", "suspended", initiator, {
      reason,
    });
    return agent;
  }

  resume(agentId: string, initiator?: Initiator): AgentRecord {
    const agent = this.getAgentOrThrow(agentId);
    this.applyTransition(agent, "resume", "active", initiator);
    return agent;
  }

  deprecate(
    agentId: string,
    reason: string = "",
    successorId: string = "",
    initiator?: Initiator,
  ): AgentRecord {
    const agent = this.getAgentOrThrow(agentId);
    this.applyTransition(agent, "deprecate", "deprecated", initiator, {
      reason,
      successor_id: successorId,
    });
    return agent;
  }

  decommission(
    agentId: string,
    reason: string = "end_of_life",
    estateDisposition?: Record<string, unknown>,
    initiator?: Initiator,
  ): AgentRecord {
    const agent = this.getAgentOrThrow(agentId);
    this.applyTransition(agent, "decommission", "decommissioned", initiator, {
      reason,
      estate_disposition: estateDisposition ?? {},
    });
    return agent;
  }

  emergencyDecommission(
    agentId: string,
    reason: string = "compromised",
    initiator?: Initiator,
  ): AgentRecord {
    const agent = this.getAgentOrThrow(agentId);
    if ((TERMINAL_STATES as readonly string[]).includes(agent.state)) {
      throw new LifecycleError(
        `Cannot emergency-decommission agent in terminal state: ${agent.state}`,
      );
    }

    const event = new LifecycleEvent({
      eventType: "emergency_decommission",
      agentId,
      agentStateBefore: agent.state,
      agentStateAfter: "decommissioned",
      initiator: initiator ?? new Initiator(),
      details: { reason, emergency: true },
    });

    this._runPreHooks(event, agent);

    agent.state = "decommissioned";
    agent.updatedAt = nowIso();
    agent.eventHistory.push(event.eventId);

    event.computeHash();
    this.store.appendEvent(event);
    this.store.saveAgent(agent);

    this._runPostHooks(event, agent);

    return agent;
  }

  fail(
    agentId: string,
    error: string = "",
    initiator?: Initiator,
  ): AgentRecord {
    const agent = this.getAgentOrThrow(agentId);
    this.applyTransition(agent, "fail", "failed", initiator, { error });
    return agent;
  }

  getAgent(agentId: string): AgentRecord | null {
    return this.store.getAgent(agentId);
  }

  getAgentOrThrow(agentId: string): AgentRecord {
    const agent = this.store.getAgent(agentId);
    if (agent === null) {
      throw new LifecycleError(`Agent not found: ${agentId}`);
    }
    return agent;
  }
}
